File: architecture/presentation/src/main/java/com/pacific/presentation/Fragment.java
package com.pacific.presentation;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;
import dagger.android.support.DaggerFragment;
import io.reactivex.internal.functions.ObjectHelper;
import javax.inject.Inject;

public abstract class Fragment extends DaggerFragment {

  @Inject
  protected ViewModelFactory modelFactory;
  private ViewModel realViewModel;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    if (isAttachViewModel()) {
      switch (modelProvider()) {
        case ACTIVITY:
          realViewModel = ViewModelProviders.of(getActivity(), modelFactory)
              .get(modelClass());
          break;
        case PARENT_FRAGMENT:
          realViewModel = ViewModelProviders.of(getParentFragment(), modelFactory)
              .get(modelClass());
          break;
        case NONE:
          realViewModel = ViewModelProviders.of(this, modelFactory)
              .get(modelClass());
          break;
        default:
          throw new UnsupportedOperationException();
      }
    }
  }

  @Override
  public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    if (!view.isClickable()) {
      view.setClickable(true);
    }
  }

  protected ViewModelSource modelProvider() {
    return ViewModelSource.ACTIVITY;
  }

  @SuppressWarnings("unchecked")
  protected final <T extends ViewModel> T fetchViewModel() {
    return (T) ObjectHelper.requireNonNull(realViewModel, "realViewModel = null");
  }

  protected boolean isAttachViewModel() {
    return true;
  }

  protected abstract Class<? extends ViewModel> modelClass();
}

File: architecture/arch-kotlin/src/main/java/com/pacific/arch/presentation/Fragment.kt
package com.pacific.arch.presentation

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.view.View
import dagger.android.support.DaggerFragment
import javax.inject.Inject

abstract class Fragment : DaggerFragment() {
    @Inject
    protected lateinit var modelFactory: ViewModelFactory

    @Suppress("UNCHECKED_CAST")
    private val realViewModel: ViewModel? by lazy {
        if (isAttachViewModel()) {
            when (modelProvider()) {
                ViewModelSource.ACTIVITY -> return@lazy ViewModelProviders.of(activity!!, modelFactory)
                        .get<ViewModel>(modelClass() as Class<ViewModel>)
                ViewModelSource.PARENT_FRAGMENT -> return@lazy ViewModelProviders.of(parentFragment!!, modelFactory)
                        .get<ViewModel>(modelClass() as Class<ViewModel>)
                ViewModelSource.NONE -> return@lazy ViewModelProviders.of(this, modelFactory)
                        .get<ViewModel>(modelClass() as Class<ViewModel>)
                else -> throw UnsupportedOperationException()
            }
        }
        null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (!view.isClickable) {
            view.isClickable = true
        }
    }

    protected open fun modelProvider(): ViewModelSource {
        return ViewModelSource.ACTIVITY
    }

    @Suppress("UNCHECKED_CAST")
    protected fun <T : ViewModel> fetchViewModel(): T = realViewModel as T

    protected open fun isAttachViewModel(): Boolean {
        return true
    }

    protected abstract fun modelClass(): Class<out ViewModel>
}